cpp/locality/LinkCell.cc

```diff
@@ -465,29 +515,35 @@ NeighborPoint LinkCellQueryIterator::next()
                     }
                 }
 
-                ++m_neigh_cell_iter;
+                while (true)
+                {
+                    ++m_neigh_cell_iter;
 
-                // In cases where we need to check the entire box, make sure
-                // that we don't check the same cell on the positive and
-                // negative sides of the IteratorCellShell cube.
-                const vec3<int> neighbor_cell_delta(*m_neigh_cell_iter);
-                if (2 * neighbor_cell_delta.x + 1 > (int) m_linkcell->getCellIndexer().getW())
-                    continue;
-                else if (2 * neighbor_cell_delta.y + 1 > (int) m_linkcell->getCellIndexer().getH())
-                    continue;
-                else if (2 * neighbor_cell_delta.z + 1 > (int) m_linkcell->getCellIndexer().getD())
-                    continue;
+                    if (m_neigh_cell_iter == IteratorCellShell(max_range, m_neighbor_query->getBox().is2D()))
+                    {
+                        break;
+                    }
+
+                    // In cases where we need to check the entire box, make sure
+                    // that we don't check the same cell on the positive and
+                    // negative sides of the IteratorCellShell cube.
+                    if (check_double_range_shell(m_linkcell, *m_neigh_cell_iter))
+                        continue;
+
+                    const unsigned int neighbor_cell = get_neighbor_cell_index(m_linkcell, point_cell, *m_neigh_cell_iter);
+
+                    auto searched_cell_iter = m_searched_cells.find(neighbor_cell);
+                    if (searched_cell_iter == m_searched_cells.end())
+                    {
+                        // This cell has not been searched yet, so we will iterate
+                        // over its contents. Otherwise, we loop back, increment
+                        // the cell shell iterator, and try the next one.
+                        m_searched_cells.insert(neighbor_cell);
+                        m_cell_iter = m_linkcell->itercell(neighbor_cell);
+                        break;
+                    }
+                }
```

Comment:
I almost want to try and move this into a function of the LinkCellIterator, but not worth trying in this PR. Just noting it so that we remember the possibility.

Reply:
There would need to be some virtual function dispatch to iterator specific termination routines, but most of the code is common.